### PR TITLE
Lesson content for Durable Functions - Advanced Patterns

### PR DESCRIPTION
**Pull request by lechnerc77**
### Content

This pull request contains the content for a lesson on advanced pattern with Durable Functions. It contains the following parts:

- Fan-out/Fan-in
- Sub-orchestration
- External Events 

The pull reqeust fixes the issue #24.

### PR Checklist

- [X] Prerequisites README.md is complete.
- [X] Lesson README.md is complete.
- [X] Source code is complete (one project with several exercises).
- **Partially**: Lesson Workspace is working (check Win+macOS): Check executed on Windows 
- [X] Link to the lesson is added to tables in Root & Lessons READMEs.
- [X] A CodeTour is available for each exercise in the lesson.
- **Partially**: The CodeTours are working in the lesson Workspace (check Win+macOS): Check executed on Windows
- [X] A Build step is added to specific GitHub workflow (if applicable).
- [X] GitHub issue is linked. Fixes: #24.

### CODE REVIEW

**Comment by marcduiker** [comment]
Thanks @lechnerc77! I'll review in the next few days and check it on macOS.

File: lessons/README.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Lesson|.NET Core|.NET 6|Typescript|PowerShell|Python|Agnostic|Contributions by
 |Deployment to Azure|-|-|-|-|[Python](python/http/http-lesson-deploy.md)|[Agnostic](deployment/deployment-lesson.md)|Marc, Dana
 |Cosmos DB Trigger & Bindings|[.NET Core](dotnetcore31/cosmosdb/cosmosdb-lesson-dotnet.md)|-|-|-|-|-|Gabriela, Marc
 |Durable Functions I |-|-|[TypeScript](typescript/durable-functions/chaining/chaining-lesson-ts.md)|-|-|-|Christian, Marc
+|Durable Functions II |-|-|[TypeScript](typescript/durable-functions/advanced/advanced-patterns-lesson.md)|-|-|-|Christian, Marc
```
Comment (marcduiker) [inline lessons/README.md:13]: Could you rename the lesson markdown to README.md? Then GH will show it by default when the folder is opened.

File: test/typescript/durable-functions/advanced-patterns/advanced-patterns.http [hunk 2]
```diff
@@ -0,0 +1,57 @@
+### Call onboarding orchestration for a single employee
```
Comment (marcduiker) [inline test/typescript/durable-functions/advanced-patterns/advanced-patterns.http:1]: I suggest to used named requests to you don't need to copy/paste guids and urls :)

```http
### 1a. Call onboarding orchestration for a single employee
# @name onboardingRequest 
POST http://localhost:7071/api/orchestrators/OnboardingOrchestrator
content-type: application/json

{
    "name": "Johnny Lawrence",
    "email": "johnny.lawrence@eaglefang.com",
    "startdate": "01.02.2022",
    "role": "developer"
}

### 1b. Fetch result from status endpoint
@id1={{onboardingRequest.response.body.$.id}}
GET http://localhost:7071/runtime/webhooks/durabletask/instances/{{id1}}

### 2a. Call onboarding orchestration for multiple employees (sub-orchestration)
# @name parallelOnboardingRequest 
POST http://localhost:7071/api/orchestrators/ParallelOnboardingOrchestrator
content-type: application/json

{
    "employees2onboard": [
        {
            "name": "Johnny Lawrence",
            "email": "johnny.lawrence@eaglefang.com",
            "startdate": "01.02.2022",
            "role": "developer"
        },
        {
            "name": "Daniel LaRusso",
            "email": "daniel.larusso@miyagido.com",
            "startdate": "15.02.2022",
            "role": "sales"
        },
        {
            "name": "Terry Silver",
            "email": "terry.silver@cobrakai.com",
            "startdate": "01.03.2022",
            "role": "manager"
        }
    ]
}

### 2b. Fetch result from status endpoint
@id2={{parallelOnboardingRequest.response.body.$.id}}
GET http://localhost:7071/runtime/webhooks/durabletask/instances/{{id2}}


### 3a. Call Orchestrator Function waiting for external event
# @name externalEventRequest
POST http://localhost:7071/api/orchestrators/OnboardingOrchestratorExternalEvent
content-type: application/json

{
    "name": "Daniel LaRusso",
    "email": "daniel.larusso@miyagido.com",
    "startdate": "15.02.2022",
    "role": "sales"
}

### 3b. Send ApprovalRequest event to endpoint
@eventName=ApprovalRequest
@id3={{externalEventRequest.response.body.$.id}}
POST http://localhost:7071/runtime/webhooks/durabletask/instances/{{id3}}/raiseEvent/{{eventName}}
content-type: application/json

 "approved"

### 3c. Fetch result from status endpoint
GET http://localhost:7071/runtime/webhooks/durabletask/instances/{{id3}}

```

File: .tours/typescript/durable-functions-advanced/.tours/02-fan-outfan-in.tour [hunk 3]
```diff
@@ -0,0 +1,231 @@
+{
+  "$schema": "https://aka.ms/codetour-schema",
+  "title": "TypeScript - Durable Functions Advanced: 02. Fan-Out/Fan-In",
```
Comment (marcduiker) [inline .tours/typescript/durable-functions-advanced/.tours/02-fan-outfan-in.tour:3]:
```suggestion
  "title": "Durable Functions Advanced: 02. Fan-Out/Fan-In",
```

File: .tours/typescript/durable-functions-advanced/.tours/03-sub-orchestration.tour [hunk 4]
```diff
@@ -0,0 +1,102 @@
+{
+  "$schema": "https://aka.ms/codetour-schema",
+  "title": "TypeScript - Durable Functions Advanced: 03. Sub-Orchestration",
```
Comment (marcduiker) [inline .tours/typescript/durable-functions-advanced/.tours/03-sub-orchestration.tour:3]:
```suggestion
  "title": "Durable Functions Advanced: 03. Sub-Orchestration",
```

File: .tours/typescript/durable-functions-advanced/.tours/04-external-events.tour [hunk 5]
```diff
@@ -0,0 +1,131 @@
+{
+  "$schema": "https://aka.ms/codetour-schema",
+  "title": "TypeScript - Durable Functions Advanced: 04. External Events - Human Interaction",
```
Comment (marcduiker) [inline .tours/typescript/durable-functions-advanced/.tours/04-external-events.tour:3]:
```suggestion
  "title": "Durable Functions Advanced: 04. External Events - Human Interaction",
```

File: .tours/typescript/durable-functions-advanced/.tours/homework/durable-functions-advanced-homework.tour [hunk 6]
```diff
@@ -0,0 +1,377 @@
+{
+  "$schema": "https://aka.ms/codetour-schema",
+  "title": "durable-functions-advanced-homework",
```
Comment (marcduiker) [inline .tours/typescript/durable-functions-advanced/.tours/homework/durable-functions-advanced-homework.tour:3]:
```suggestion
  "title": "Durable Functions Advanced: Homework",
```

File: .tours/typescript/durable-functions-advanced/.tours/homework/durable-functions-advanced-homework.tour [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    },
+    {
+      "title": "Challenge 1 (Fan-in/Fan-Out) - Setup",
+      "description": "Create a new Function App via the Azure Functions extension of VSCode with the following settings:\r\n\r\n1. Name the project `BusinessTripBookingByDurableFunction`.\r\n2. Choose `TypeScript` as language.\r\n3. Select `Durable Functions HTTP Starter` as a template as we want to trigger the execution of the Durable Function via an HTTP call.\r\n4. Name the function `BusinessTripBookingStarter`.\r\n5. Set the authorization level of the function to `Anonymous`."
+    },
+    {
+      "title": "Challenge 1 (Fan-in/Fan-Out)  - Install dependencies",
+      "description": "Install the npm dependencies via:\r\n\r\n>> npm install durable-functions\r\n>> npm install @types/node --save-dev"
```
Comment (marcduiker) [inline .tours/typescript/durable-functions-advanced/.tours/homework/durable-functions-advanced-homework.tour:15]:
```suggestion
      "description": "Install the npm dependencies via:\r\n\r\n>> npm install durable-functions\r\n\r\n>> npm install @types/node --save-dev"
```

**Review by marcduiker — CHANGES_REQUESTED** [review]
Oof this PR is big 😅 . 

Great contribution again @lechnerc77 !

There are some minor suggestions here and there. Can you also check if all the npm dependencies are up to date? It took me a bit too long to look at this so I think there are some vulnerabilities now 😬 . 

Looking forward to host a Durable Functions session with you! 😄

**Comment by marcduiker** [comment]
Workspaces & CodeTours work on macOS btw 👍

File: lessons/README.md (hunk 1 above)
Comment (lechnerc77) [reply lessons/README.md:13]: Will do so

Comment (lechnerc77) [reply lessons/README.md:14]: Done

**Comment by lechnerc77** [comment]
npm audit vulnerabilities are fixed

File: test/typescript/durable-functions/advanced-patterns/advanced-patterns.http (hunk 2 above)
Comment (lechnerc77) [reply test/typescript/durable-functions/advanced-patterns/advanced-patterns.http:1]: Done

**Review by marcduiker — APPROVED** [review]
Thanks for the changes @lechnerc77, looks in great shape! 💪

**Pull request merged**